### DISCUSSION TITLE
FFmpeg HEVC FLV

### DISCUSSION THREAD

**Discussion by examedia-video-engineering** [shogo4405/HaishinKit.kt#297] [Q&A]
Greetings!
can't seem to recieve and transcode a HEVC FLV stream from iOS, get contant erors from FFmpeg 7.0.2 (the latest for now). Apperently there is code  in it already to support HEVC in FLV but its not working. Is there a specific patch that might enable FFmpeg 7.x to parse and decode RTMP FLV streams with HEVC codec from iOS HaisinKit ?

**Comment by shogo4405** [comment]
This discussion is for Android, but is it okay to talk about iOS?

If it's for iOS, I have confirmed that it works on my end, so I would like to know the steps to reproduce the issue. For Android, I have not tested the HaishinKit for iOS, so if it doesn't work, that might be the reason.

The conditions for use are:
- The server you are using needs to support Enhanced RTMP.
- It does not support the independently extended HEVC before Enhanced RTMP.

**Reply by examedia-video-engineering** [reply]
Sorry, lets keep it Android related. 
The source is Android RTMP with HEVC video and AAC audio now.

FFmpeg version 7.0.2, which has native support of Enhanced RTMP, fails to decode with errors like: 

`ffmpeg -i rtmp://myserver/streamX -f null -`

error messages are: 
`
[aac @ 0x7fa8ec2cca80] This decoder does not support parameter changes, but PARAM_CHANGE side data was sent to it.
[aac @ 0x7fa8ec2cca80] Error applying parameter changes.
[flv @ 0x7fa8ec32bd80] Video codec (c) is not implemented. Update your FFmpeg version to the newest one from Git. If the problem still occurs, it means that your file has a feature which has not been implemented.
[flv @ 0x7fa8ec32bd80] If you want to help, upload a sample of this file to https://streams.videolan.org/upload/ and contact the ffmpeg-devel mailing list. (ffmpeg-devel@ffmpeg.org)
[flv @ 0x7fa8ec32bd80] Video codec (c) is not implemented. Update your FFmpeg version to the newest one from Git. If the problem still occurs, it means that your file has a feature which has not been implemented.
`

Any advice on processing FLV HEVC stream with FFmpeg please?

**Reply by shogo4405** [reply]
It works me.
```
trunk % ffplay rtmp://localhost/live/live
ffplay version 6.1.1 Copyright (c) 2003-2023 the FFmpeg developers
  built with Apple clang version 15.0.0 (clang-1500.3.9.4)
  configuration: --prefix=/opt/homebrew/Cellar/ffmpeg/6.1.1_6 --enable-shared --enable-pthreads --enable-version3 --cc=clang --host-cflags= --host-ldflags='-Wl,-ld_classic' --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libaribb24 --enable-libbluray --enable-libdav1d --enable-libharfbuzz --enable-libjxl --enable-libmp3lame --enable-libopus --enable-librav1e --enable-librist --enable-librubberband --enable-libsnappy --enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libtesseract --enable-libtheora --enable-libvidstab --enable-libvmaf --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-lzma --enable-libfontconfig --enable-libfreetype --enable-frei0r --enable-libass --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libopenvino --enable-libspeex --enable-libsoxr --enable-libzmq --enable-libzimg --disable-libjack --disable-indev=jack --enable-videotoolbox --enable-audiotoolbox --enable-neon
  libavutil      58. 29.100 / 58. 29.100
  libavcodec     60. 31.102 / 60. 31.102
  libavformat    60. 16.100 / 60. 16.100
  libavdevice    60.  3.100 / 60.  3.100
  libavfilter     9. 12.100 /  9. 12.100
  libswscale      7.  5.100 /  7.  5.100
  libswresample   4. 12.100 /  4. 12.100
  libpostproc    57.  3.100 / 57.  3.100
Input #0, flv, from 'rtmp://localhost/live/live': sq=    0B f=0/0   
  Metadata:
    |RtmpSampleAccess: true
    server          : SRS/7.0.0(Hang)
    server_version  : 7.0.0
  Duration: N/A, start: 0.000000, bitrate: 720 kb/s
  Stream #0:0: Data: none
  Stream #0:1: Audio: aac (LC), 44100 Hz, mono, fltp, 65 kb/s
  Stream #0:2: Video: hevc (Main), yuv420p(tv, bt709), 854x480, 655 kb/s, 30 fps, 30 tbr, 1k tbn
  33.61 A-V: -0.002 fd=   0 aq=    9KB vq=   81KB sq=    0B f=0/0   
```

<img width="843" alt="スクリーンショット 2024-08-10 17 11 10" src="https://github.com/user-attachments/assets/4ec8b246-4c13-4df9-945b-65036b31a9be">

**Reply by shogo4405** [reply]
I update ffmpeg version 7.0.2. And works for me.
- With Enhanced RTMP, does the server side also need to be compatible? Is it compatible?

**Reply by examedia-video-engineering** [reply]
Interesting! 
Thank you very much, it seems the problem is on my side.

**Comment by examedia-video-engineering** [comment]
Basically question is: "Would you consider modifying FLV format to align with Veovera's https://github.com/veovera/enhanced-rtmp/tree/main ? Which seems to align with YuoTube, FB and others." 
Thank you.

**Reply by shogo4405** [reply]
What does FLV refer to?

**Reply by examedia-video-engineering** [reply]
FLV is Adobe Flash Video, the general muxer for RTMP. Very old format, but still works fine as it seems ))
